File: samples_harder/sample.py
def nested_try(a, b):
    result = 0
    try:
        try:
            if b == 0:
                raise ZeroDivisionError("zero")
            result = a // b
        except ZeroDivisionError:
            if a > 0:
                result = a
            else:
                raise ValueError("both bad")
        result += 1
    except ValueError:
        result = -999
    finally:
        print("outer finally")
    return result


def try_in_loop(items):
    results = []
    for item in items:
        try:
            if item is None:
                raise TypeError("null item")
            if item == 0:
                raise ValueError("zero item")
            results.append(100 // item)
        except TypeError:
            continue
        except ValueError:
            results.append(-1)
            if len(results) > 5:
                break
        finally:
            print("processed")
    return results


def deep_match(data):
    output = []
    for item in data:
        match item:
            case {"type": "a", "val": v}:
                if v > 0:
                    match v % 3:
                        case 0:
                            output.append("a-triple")
                        case 1:
                            output.append("a-rem1")
                        case _:
                            output.append("a-other")
                else:
                    output.append("a-neg")
            case {"type": "b", "val": v}:
                if v == 0:
                    continue
                output.append("b")
            case _:
                break
    return output


def complex_while_for(matrix):
    total = 0
    row = 0
    while row < len(matrix):
        if not matrix[row]:
            row += 1
            continue
        for col, val in enumerate(matrix[row]):
            if val is None:
                break
            if val < 0:
                continue
            if val == 0:
                total += col
            elif val > 100:
                return total
            else:
                total += val
        else:
            row += 1
            continue
        row += 2
    return total


def multi_except_finally(path, mode):
    result = None
    try:
        with open(path) as f:
            data = f.read()
            if not data:
                raise IOError("empty")
            match mode:
                case "upper":
                    result = data.upper()
                case "lower":
                    result = data.lower()
                case _:
                    raise ValueError("bad mode")
    except IOError:
        result = "io_error"
    except ValueError:
        result = "val_error"
    except Exception:
        raise
    finally:
        print("cleanup")
    return result


def generator_with_control(items):
    for i, item in enumerate(items):
        if item is None:
            continue
        if item < 0:
            return
        if item == 0:
            yield i
        else:
            for sub in range(item):
                if sub > 3:
                    break
                yield sub


def recursive_descent(tokens, pos):
    if pos >= len(tokens):
        return None, pos

    match tokens[pos]:
        case "(":
            inner, new_pos = recursive_descent(tokens, pos + 1)
            if new_pos < len(tokens) and tokens[new_pos] == ")":
                return inner, new_pos + 1
            else:
                raise SyntaxError("missing )")
        case ")":
            return None, pos
        case _:
            return tokens[pos], pos + 1


def main():
    nested_try(10, 0)
    nested_try(-5, 0)
    try_in_loop([1, None, 0, 2, None, 0, 3])
    deep_match([
        {"type": "a", "val": 9},
        {"type": "a", "val": -1},
        {"type": "b", "val": 0},
        {"type": "b", "val": 5},
        42,
    ])
    complex_while_for([[1, 2, None], [], [3, -1, 4]])
    list(generator_with_control([0, 2, -1, 5]))
    recursive_descent(["(", "x", ")"], 0)

File: samples_harder/sample.java
class SampleHarder {

    int value;

    SampleHarder(int v) {
        if (v < 0) {
            throw new IllegalArgumentException("negative");
        }
        value = v;
    }

    int nestedTryCatch(int a, int b) {
        int result;
        try {
            try {
                if (b == 0) {
                    throw new ArithmeticException("zero");
                }
                result = a / b;
            } catch (ArithmeticException inner) {
                if (a > 0) {
                    result = a;
                } else {
                    throw new RuntimeException("both bad");
                }
            }
            result += 1;
        } catch (RuntimeException outer) {
            result = -999;
        } finally {
            System.out.println("outer finally");
        }
        return result;
    }

    int tryInLoop(int[] items) {
        int sum = 0;
        for (int item : items) {
            try {
                if (item < 0) {
                    throw new IllegalArgumentException("negative");
                }
                sum += 100 / item;
            } catch (ArithmeticException e) {
                sum += -1;
                if (sum < -10) {
                    break;
                }
            } catch (IllegalArgumentException e) {
                continue;
            } finally {
                System.out.println("processed");
            }
        }
        return sum;
    }

    int switchWithLoops(int[] data) {
        int state = 0;
        int output = 0;

        for (int i = 0; i < data.length; i++) {
            switch (state) {
                case 0:
                    if (data[i] > 0) {
                        state = 1;
                    } else if (data[i] < 0) {
                        state = 2;
                    } else {
                        continue;
                    }
                    output++;
                    break;
                case 1:
                    for (int j = 0; j < data[i]; j++) {
                        output += j;
                        if (output > 100) {
                            return output;
                        }
                    }
                    state = 0;
                    break;
                case 2:
                    if (data[i] == -99) {
                        return -1;
                    }
                    output -= data[i];
                    state = 0;
                    break;
            }
        }
        return output;
    }

    int labeledBreaks(int[][] matrix) {
        int found = -1;
        outer:
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] < 0) {
                    continue;
                }
                if (matrix[i][j] == 0) {
                    found = i;
                    break;
                }
                if (matrix[i][j] == 99) {
                    found = i * 100 + j;
                    break outer;
                }
            }
            if (found >= 0) {
                break;
            }
        }
        return found;
    }

    String deepNesting(int[] items) {
        StringBuilder sb = new StringBuilder();
        int i = 0;
        while (i < items.length) {
            if (items[i] > 0) {
                for (int j = 0; j < items[i]; j++) {
                    if (j % 2 == 0) {
                        try {
                            if (items[i] / (j + 1) > 10) {
                                throw new ArithmeticException("too big");
                            }
                            sb.append(items[i] / (j + 1));
                        } catch (ArithmeticException e) {
                            sb.append("E");
                            continue;
                        }
                    } else {
                        sb.append("-");
                    }
                }
            } else if (items[i] == 0) {
                i++;
                continue;
            } else {
                return sb.toString();
            }
            i++;
        }
        return sb.toString();
    }

    int switchFallthrough(int code) {
        int result = 0;
        switch (code) {
            case 1:
            case 2:
            case 3:
                result = code * 10;
                break;
            case 4:
                result = 40;
            case 5:
                result += 50;
                break;
            case 6:
                if (code > 5) {
                    result = 60;
                    break;
                }
            case 7:
                result = 70;
                break;
            default:
                result = -1;
                break;
        }
        return result;
    }

    void doWhileNested(int limit) {
        int i = 0;
        do {
            int j = 0;
            while (j < limit) {
                if (i == j) {
                    j++;
                    continue;
                }
                if (i + j > limit * 2) {
                    break;
                }
                j++;
            }
            i++;
            if (i == 3) {
                continue;
            }
        } while (i < limit);
    }

    void run() {
        nestedTryCatch(10, 0);
        nestedTryCatch(-5, 0);
        int[] items = {1, 0, -1, 2, 3};
        tryInLoop(items);
        switchWithLoops(items);
        int[][] matrix = {{1, 2, 99}, {-1, 0, 3}};
        labeledBreaks(matrix);
        deepNesting(items);
        switchFallthrough(4);
        doWhileNested(4);
    }
}
